### ISSUE TITLE
Caffe protobuf crash - OSX

### ISSUE THREAD

**Issue by orgicus** [CMU-Perceptual-Computing-Lab/openpose#168]
### Issue summary

Running openpose.bin crashes. Errors seem related to caffee.proto:

```bash
./openpose.bin --video ../../../examples/media/video.avi 
objc[40481]: Class CaptureDelegate is implemented in both /usr/local/opt/opencv/lib/libopencv_highgui.2.4.dylib and /Users/George/Downloads/Grouped/cv/opencv/build/install/lib//libopencv_videoio.3.2.dylib. One of the two will be used. Which one is undefined.
[libprotobuf ERROR google/protobuf/descriptor_database.cc:57] File already exists in database: caffe.proto
[libprotobuf FATAL google/protobuf/descriptor.cc:1018] CHECK failed: generated_database_->Add(encoded_file_descriptor, size): 
libc++abi.dylib: terminating with uncaught exception of type google::protobuf::FatalException: CHECK failed: generated_database_->Add(encoded_file_descriptor, size): 
Abort trap: 6
```

What should I double check or how should I setup the project to run it correctly ?

### Executed command (if any)
`./openpose.bin --video ../../../examples/media/video.avi `
Adding `--logging_level 0` didn't return any extra debug information.

### Type of issue
You might select multiple topics, delete the rest:
- Execution error
- Help wanted
- Question

### Your system configuration
**Operating system** OSX 10.11.5
**CUDA version** CUDA Driver Version: 8.0.61, CUDA SDK 8.0
**cuDNN version**:5.1
**GPU model** NVIDIA GeForce GT 750M 2048 MB
**Caffe version**: Default from OpenPose
**OpenCV version**: OpenCV 3.2 build from source (with GPU support)
**Generation mode (only for Ubuntu): [Modified Makefile](https://github.com/orgicus/openpose/blob/master/Makefile) + [Makefile.config.OSX_10_5.example](https://github.com/orgicus/openpose/blob/master/Makefile.config.OSX_10_5.example)
Compiler (`gcc --version` in Ubuntu): ```bash
Configured with: --prefix=/Applications/IDEsAndEditors/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1 Apple LLVM version 7.0.0 (clang-700.0.72)```

**Comment by orgicus** [comment]
After looking into the issue further there may be a case of the protobuf library being linked and loaded twice as far as I can tell from [this answer](https://stackoverflow.com/questions/33017985/static-linking-with-generated-protobufs-causes-abort).

Initially I have compiled OpenCV from source with CUDA support and the contrib modules. The dnn related module (`libopencv_dnn_modern.3.2.0.dylib`) seem to already link libprotobuf. 
I've compiled OpenCV from source again, this time disabling the dnn module. 

I'm a bit confused at the moment: is the dnn OpenCV module required ?
Is there a particular OpenCV configuration required ?

**Comment by gineshidalgo99** [comment]
Default OpenCV is fine, and you should be able to disable the DNN module. The old 2.4.9 version is usually the easiest to install and use, 3.X is perfectly fine but it usually gives more problems in Ubuntu

**Comment by orgicus** [comment]
@gineshidalgo99 Thank you for the tips.
I've made a bit of progress but then got stuck again.

I had OpenCV 2.4.13 installed via homebrew with no GPU support which caused a lot of errors when compiling caffee with GPU support. I've downloaded and compiled OpenCV 2.4.13 from source with GPU support, libcaffee compiled, openpose.bin compiled, but for some reason it can't seem to find the opencv libraries.

I've tried using [install_name_tool](http://thecourtsofchaos.com/2013/09/16/how-to-copy-and-relink-binaries-on-osx/) and set an absolute path to the opencv libraries:
```bash
otool -L ./build/examples/openpose/openpose.bin 
./build/examples/openpose/openpose.bin:
	@rpath/libopenpose.so.1.0.0-rc3 (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libcaffe.so.1.0.0 (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libcudart.8.0.dylib (compatibility version 0.0.0, current version 8.0.61)
	@rpath/libcublas.8.0.dylib (compatibility version 0.0.0, current version 8.0.61)
	@rpath/libcurand.8.0.dylib (compatibility version 0.0.0, current version 8.0.61)
	/usr/local/opt/glog/lib/libglog.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/local/opt/gflags/lib/libgflags.2.dylib (compatibility version 2.0.0, current version 2.1.2)
	/usr/local/opt/boost/lib/libboost_system.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/boost/lib/libboost_filesystem.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
	/usr/local/opt/hdf5/lib/libhdf5_hl.100.dylib (compatibility version 101.0.0, current version 101.1.0)
	/usr/local/opt/hdf5/lib/libhdf5.101.dylib (compatibility version 102.0.0, current version 102.0.0)
	/usr/local/opt/leveldb/lib/libleveldb.1.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/snappy/lib/libsnappy.1.dylib (compatibility version 5.0.0, current version 5.0.0)
	/usr/local/opt/lmdb/lib/liblmdb.dylib (compatibility version 0.0.0, current version 0.0.0)
	/Users/George/Downloads/Grouped/cv/opencv-2.4.13/buildOpenPoseGPU/install/lib/libopencv_core.2.4.dylib (compatibility version 2.4.0, current version 2.4.13)
	/Users/George/Downloads/Grouped/cv/opencv-2.4.13/buildOpenPoseGPU/install/lib/libopencv_highgui.2.4.dylib (compatibility version 2.4.0, current version 2.4.13)
	/Users/George/Downloads/Grouped/cv/opencv-2.4.13/buildOpenPoseGPU/install/lib/libopencv_imgproc.2.4.dylib (compatibility version 2.4.0, current version 2.4.13)
	/Users/George/Downloads/Grouped/cv/opencv-2.4.13/buildOpenPoseGPU/install/lib/libopencv_contrib.2.4.dylib (compatibility version 2.4.0, current version 2.4.13)
	/usr/local/opt/boost/lib/libboost_thread-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libcudnn.5.dylib (compatibility version 0.0.0, current version 5.1.10)
	/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.1.0)
```
However the error persists:
```bash
./build/examples/openpose/openpose.bin 
dyld: Library not loaded: libopencv_core.2.4.dylib
  Referenced from: /Users/George/Downloads/Grouped/code/openpose/.build_release/lib/libopenpose.so.1.0.0-rc3
  Reason: image not found
Trace/BPT trap: 5
```

**Comment by orgicus** [comment]
Woo hoo! 
I've managed to get finally get it to run on OSX, but in a very hacky way at the moment:

1. I've used `install_name_tool` to change the opencv paths in `libcaffe`,`libopenpose` and `openpose.bin`
2. Symlinked opencv's libs to openpose's folder

My guess is I've botched the Makefiles for both libcaffe and openpose and will need to spend more time to get the flags and paths right.

@gineshidalgo99 Thank you so much for pointing me towards OpenCV 2 instead of 3, pretty sure you saved me a hefty headache.

I'm only getting 1.9fps with `--net_resolution 320x176 --resolution 640x480 --camera_resolution 640x480` at the moment but I'll need to play with all the flags a bit more. 

Any tips on getting the speed are appreciated.
